[web] Support libevent as WS server instead of libwebsockets

chme opened the pull request:
The next planned minor release 2.2 of libevent will have support for running a websocket server. I thought, I give it a try and see if it works in OwnTone and check if it is something worth adding :-)

The changes in this PR add a libevent based websocket server, that accepts connections under the path `/ws` (and the same port as the web interface). 

If libevent >= 2.2 is detected during configure and "websocket_port" == 0 in the config file, the libwebsocket implementation is disabled and instead the libevent websocket server is active.

The web interface was updated to attempt a WS connection to `/ws` if websocket_port == 0.

The advantages in using libevent are - if it proves to be reliable - that the libwebsocket dependency could be removed. Additionally, only having one port for the web interface, will make it easier to use a reverse proxy with OwnTone (especially if the reverse proxy is running on the same host). 

I did some short tests, and it seems to be working fine, but more tests are required. 
I tested it locally and by building an OwnTone container with libevent from git-master (<https://github.com/chme/owntone-container/tree/feat/libevent-git>). Also I am unsure how it will work with the multi threading support of the http server in OwnTone.

Currently only an alpha release of libevent 2.2 is available (<https://github.com/libevent/libevent/releases/tag/release-2.2.1-alpha>) . Note, that the libevent 2.2.1 alpha release contains a bug that prevents ws connections from Firefox (Chrome is working fine). The bug is already fixed in the master branch of libevent.

ejurgensen commented:
This looks great, I think it would be very good to have an alternative to libwebsockets. Maybe this can also be a way of solving issue #1699 and the problem mentioned in #1446.

I will try to check the http multithreading to see if this could cause any problems.

hacketiwack commented:
Sounds really promising. Too bad that the libevent team is slow at releasing versions of their library; last commit was last year.

chme commented:
> Too bad that the libevent team is slow at releasing versions of their library; last commit was last year.

Yes, more frequent releases of libevent would be nice. Especially because it also takes quite a long time until all the distros pick up the new release.

ejurgensen reviewed (commented):
Here's a few comments and suggestions. I think it should work with the multithreaded http server, at least as long as THREADPOOL_NTHREADS is 1, because then there is still only one httpd thread and one evhttp instance.

chme commented on configure.ac:
```diff
@@ -143,7 +143,13 @@ OWNTONE_MODULES_CHECK([COMMON], [SQLITE3], [sqlite3 >= 3.5.0],
-	[event_base_new], [event2/event.h])
+	[event_base_new], [event2/event.h],
+	[dnl check for version 2.2 (with websocket server support)
+	 PKG_CHECK_EXISTS([libevent >= 2.2.1],
```
Might be possible ... if you want, I can look into it.

But there is now an additional breaking change with libevent >= 2.2. The proxy handling was changed and the flag `EVHTTP_PROXY_REQUEST` was removed (<https://github.com/libevent/libevent/pull/1736>).

Maybe better to check for the version? Will make it slightly easier to spot when these conditionals could be removed ...

chme commented on src/websocket_ev.c:
```diff
@@ -0,0 +1,331 @@
+  return json_response;
+}
+
+/* Thread: library (the thread the event occurred) */
```
Updated the comment (also in `websocket.c`)

chme commented on src/websocket_ev.c:
```diff
@@ -0,0 +1,331 @@
+  struct client *client = NULL;
+  char *reply = NULL;
+
+  for (client = clients; client; client = clients->next)
```
Yes, thread safety is an issue. I reworked it. The listener callback now triggers a command  executed in the httpd thread (hope I did it correctly). Required a bit of additional refactoring of the listener call back function.

The listener now supports passing an opaque pointer (context) when adding a listener. This pointer is then passed to each notify callback invocation.

chme commented on src/websocket_ev.c:
```diff
@@ -0,0 +1,331 @@
+      return -1;
+    }
+
+  DPRINTF(E_DBG, L_WEB, "notify callback request: %s\n", json_object_to_json_string(request));
```
This log is only written once per websocket connection. It should not be that noisy.
I have reduced the log level for the "notify callback event received" log message to "spam". That seems to me noisier and not the helpful.

chme commented on src/websocket_ev.c:
```diff
@@ -0,0 +1,331 @@
+
+  evhttp_set_cb(evhttp, "/ws", gencb_ws, NULL);
+
+  listener_add(listener_cb, LISTENER_UPDATE | LISTENER_DATABASE | LISTENER_PAIRING | LISTENER_SPOTIFY | LISTENER_LASTFM
```
Yes, that sounds like a good idea. Together with your suggestion to introduce a map for the event names, this would allow to implement a generic forwarding of events through the ws connection.

But ... I would like to keep it out of this PR as it is not directly related - if that is OK with you.

chme commented on src/httpd_libevhttp.c:
```diff
@@ -38,6 +38,7 @@
 #include "logger.h"
 #include "commands.h"
 #include "httpd_internal.h"
+#include "websocket_ev.h"
```
Mmh .. generally I am not a fan of huge files, but given that with the changes to make it thread safe it is more tightly coupled with `httpd_libevent.c`, I moved the code now into `httpd_libevent.c`.

chme commented on src/websocket_ev.c:
```diff
@@ -0,0 +1,331 @@
+  DPRINTF(E_DBG, L_WEB, "notify callback reply: %d\n", events);
+
+  notify = json_object_new_array();
+  if (events & LISTENER_UPDATE)
```
Yes, I agree. But in my opinion, it would be better to do it in a dedicated PR.

chme commented:
Thanks for taking the time to review this change and apologies for the lack of response from me. 

I finally got around and rebased the changes on top of `master` and addressed the review comments. The thread safety issue required some rework (that now also effects a tiny bit other code parts).

I'll test this more over the coming weeks to see if it is stable (at least for my set up).

ejurgensen commented:
Excellent, you are welcome to merge it when you think it's ready. Looking forward to using this as the default, but I guess we have to wait for a libevent release.

It's fine to keep the version check of libevent instead of feature check, and I agree the map can wait.

chme commented:
I have merged it now. It is working stable for me for quite some time now (with libevent build from git).

Too bad, there is no progress on a new libevent release though. And given how long Debian release cycles are, it will probably take around three years, until this will be usable on Debian distros ...
